nni/compression/pytorch/speedup/compress_modules.py

```diff
@@ -3,196 +3,414 @@
+    """
+    Convert the mask tensor to the coarse-grained mask tensor.
+    Parameters
+    ---------
+    t_mask: torch.Tensor
+        The tensor only have 1s and 0s, 0 indicates this value is masked
+        and 1 indicates the corresponding value is not masked.
+    dim: int
+        Try to reduce the mask tensor on this dimension.
+
+    Returns
+    -------
+    indexes: torch.Tensor
+        The indexes of the sparsity that can be structurally removed.
+    remained_indexes: torch.Tensor
+        The indexes of values that need to be remained.
+    """
+    assert isinstance(t_mask, torch.Tensor)
+    shape = list(t_mask.size())
+    n_dims = len(shape)
+    dim_list = list(range(n_dims))
+    # try to reduce the mask from the 0-th dimension
+    dim_list.remove(dim)
+
+    t_merged = torch.sum(t_mask, dim_list)
+    assert t_merged.size(0) == shape[dim]
+    all_pruned = t_merged == 0
+    need_remain = t_merged != 0
+    # return the indexes of the sparsity that can be removed
+    indexes = torch.nonzero(all_pruned, as_tuple=True)[0]
+    remained_indexes = torch.nonzero(need_remain, as_tuple=True)[0]
+    return indexes, remained_indexes
+
+
+def no_replace(module, auto_infer):
     """
     No need to replace
     """
     _logger.debug("no need to replace")
     return module
 
 
-def replace_linear(linear, mask):
+def replace_linear(linear, auto_infer):
     """
     Parameters
     ----------
     linear : torch.nn.Linear
         The linear module to be replace
-    mask : ModuleMasks
-        The masks of this module
+    auto_infer : AutoMaskInference
+        The auto mask inference object that contains the input,
+        parameter and output masks.
 
     Returns
     -------
     torch.nn.Linear
         The new linear module
     """
-    assert isinstance(mask, ModuleMasks)
-    assert mask.input_mask is not None
-    assert mask.output_mask is None
-    assert not mask.param_masks
-    index = mask.input_mask.mask_index[-1]
-    in_features = index.size()[0]
-    _logger.debug("replace linear with new in_features: %d", in_features)
-    new_linear = torch.nn.Linear(in_features=in_features,
-                                 out_features=linear.out_features,
-                                 bias=linear.bias is not None)
+    NEED_FOLD_BIAS = auto_infer.fold_bias
+    assert isinstance(linear, nn.Linear)
+    assert len(auto_infer.in_masks) == 1
+    assert isinstance(auto_infer.output_mask, torch.Tensor)
+    in_mask = auto_infer.in_masks[0]
+    # only need the first batch of the constant
+    in_constant = auto_infer.in_constants[0][:1]
+    output_mask = auto_infer.output_mask
+    weight_mask = auto_infer.weight_mask['weight']
+    # N C K
+    pruned_in, remained_in = convert_to_coarse_mask(in_mask, 1)
+    pruned_out, remained_out = convert_to_coarse_mask(output_mask, 1)
+    n_remained_in = weight_mask.size(1) - pruned_in.size(0)
+    n_remained_out = weight_mask.size(0) - pruned_out.size(0)
```

Comment:
If i got it right, `n_remained_in == remained_in.size(0)`?

Reply:
Yes, indeed.